### ISSUE TITLE
Improving quality of rough environmental reflections

### ISSUE THREAD

**Issue by donmccurdy** [mrdoob/three.js#26796]
### Description

I've attached a model using metalness=100% and roughness=15%, and compared screenshots in several engines.

[model_slice_clean.glb.zip](https://github.com/mrdoob/three.js/files/12648344/model_slice_clean.glb.zip)

**Khronos Sample Viewer**
![slice_khronos](https://github.com/mrdoob/three.js/assets/1848368/4c21dfc9-1fdd-4e2e-adaf-b2ce903f46a2)
**PlayCanvas**
![slice_playcanvas](https://github.com/mrdoob/three.js/assets/1848368/80485939-6ac9-4a84-bd94-64bd9775f3e7)
**Babylon.js**
![slice_babylon](https://github.com/mrdoob/three.js/assets/1848368/ec52910e-a0cd-4999-adfe-cfd9dec34403)
**three.js**
![slice_threejs](https://github.com/mrdoob/three.js/assets/1848368/46f0ed55-d0e4-4fff-87fd-4465a3104201)

You'll likely notice first that in Khronos Sample Viewer and PlayCanvas, the surface appears considerably less "rough". I'm not especially concerned about that here – the glTF specification doesn't include environment maps today, and implementations have some leeway in how to preprocess them.

I am wondering about the visible blocky edges in the three.js render, though. The same issue is visible with other environments, particularly RoomEnvironment. I understand we use a low-res 256x256px cubemap for PMREM by default — I tried increasing that resolution in local builds, but didn't see a meaningful reduction of the problem.

### Solution

Are there settings available, or settings we could add, to improve the rough reflections for cases like these? Can/should IBL be pre-processed offline to improve the result? For scenes with large reflective surfaces, I think the current reflections may be less clean than users would prefer.

### Alternatives

N/A

**Comment by elalish** [comment]
I can definitely answer any of your questions here in detail, but first I'd like to take issue with this: 

>You'll likely notice first that in Khronos Sample Viewer and PlayCanvas, the surface appears considerably less "rough". I'm not especially concerned about that here – the glTF specification doesn't include environment maps today, and implementations have some leeway in how to preprocess them.

The glTF spec may not include environment maps, but that's only a means of data ingest - it *does* define our BRDF, which says how roughness should affect sampling of *any* light source. To me this shows a very clear (and serious) bug on the part of the sample viewer and PlayCanvas, as they are not representing surface roughness accurately at all, which is probably the single most important part of PBR. 

As to the much smaller error that three.js exhibits from HDR pixel interpolation, agreed that it's not ideal. My guess is that three.js is doing its interpolation in HDR, while Babylon is probably doing it in SDR (due to clamping or tone mapping their lighting). I don't believe they've ever fixed their *even more serious* bug, where they clamp their input HDR environment maps, thus using only the SDR part of their range. This gives them wildly incorrect results for realistic HDRs that include e.g. the sun. 

The fact is that pixel-level interpolation of HDR data doesn't work well, but I don't know of a reasonable alternative. Really all pixel-interpolation should happen after tone mapping, but rasterizers don't really make that possible. I would say of all these renders, three.js is by far the most accurate, despite this issue. Certainly it's possible to change the mapping of roughness -> PMREM resolution. Increasing that will smooth this out, but you hit diminishing returns rapidly, as making the surface a little flatter dramatically magnifies the effect. And of course a huge part of frame rate performance is the size of the PMREM texture cache.

**Comment by donmccurdy** [comment]
We are at roughness=0.15 here — I guess I don't have a good mental model of how 15% rough "should" look. If you're confident that we're correctly representing roughness and the other two renders are not, I'm happy with that. 

Also for comparison, here is [MetalRoughSpheres](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/MetalRoughSpheres) —

**Khronos Sample Viewer**
![Screenshot 2023-09-18 at 5 20 16 PM](https://github.com/mrdoob/three.js/assets/1848368/edd32c44-06aa-4e86-afd9-0454a640d88e)


**threejs**
![Screenshot 2023-09-18 at 5 20 36 PM](https://github.com/mrdoob/three.js/assets/1848368/3d02acdc-57e3-4714-858e-c13fc83a0ecf)

***

I agree the Babylon.js render appears to not be color managed correctly. I've included it for the roughness comparison — they seem to get a similarly-rough result without the blockiness, but if that's done by clamping or tone-mapping early then that's not going to help us much I suppose... 🤔

***

> Certainly it's possible to change the mapping of roughness -> PMREM resolution ... but you hit diminishing returns rapidly ...

Even if it's a "change X and recompile three.js" situation, I'd be interested how to test this?

**Comment by elalish** [comment]
It's not parameterized for that super cleanly at the moment. I think you'd need to change this [relationship](https://github.com/mrdoob/three.js/blob/1e673dbaa2cf21370dcfc36e72a3414d65235190/src/extras/PMREMGenerator.js#L575) and make a corresponding change [here](https://github.com/mrdoob/three.js/blob/1e673dbaa2cf21370dcfc36e72a3414d65235190/src/renderers/shaders/ShaderChunk/cube_uv_reflection_fragment.glsl.js#L86).

**Comment by donmccurdy** [comment]
Apologies if this is something where I should just go read some manuals. Please feel free to shoo me away if so. :) 

I pulled a low-res snapshot of the processed IBL out of spector.js...

![download](https://github.com/mrdoob/three.js/assets/1848368/929c8096-c22a-41ae-a2f1-39285b9e4bbe)

... and now I guess I have a more basic question. Why do lower-roughness values correspond to lower-resolution areas of the texture at all? Instead of 8 cubemaps with power-of-two resolution differences, could this (hypothetically) be 8 cubemaps all at 6x256x256px? I'd expect that to have a much higher preprocessing cost, but not much difference in the final texture size and framerate?

**Comment by elalish** [comment]
Lower roughness are higher resolution because smoothing and resolution are fundamentally related - we're following the same mathematics as mipmaps. This hits two birds with one stone by also suppressing aliasing mipmap-style. With your approach you'd likely need mipmaps for each roughness level (a common approach before mine). 

Yes, the higher preprocessing cost is a very big deal - the main purpose of this technique when I wrote it was to make PMREM generation online instead of offline. Also, the unused texture portions don't actually cost you much of anything in your texture cache, since those parts don't get loaded out of memory, so it still helps with framerate as well.

**Comment by donmccurdy** [comment]
Mipmaps are traditionally selected for sampling so that the space between samples is approximately 1 texel. When a surface is viewed directly and at full scale, a high resolution mip is sampled. But we don't quite have that same connection between screen space and texel space here.

Borrowing this chart from your paper...

![Screenshot 2023-09-19 at 10 58 36 PM](https://github.com/mrdoob/three.js/assets/1848368/5e0ba29d-0c07-47e9-9d3d-9dd3b9c23823)

...it seems like perhaps this particular choice of roughness (r=0.15) has landed in a place where the mip resolution has dropped off quickly to around 32x32, but the surface is still smooth enough to show recognizable reflections. 

_Context: This also happens to be a prominent visual on a page — there are no other relevant roughness levels! — so the rest of our envmap is unfortunately going to waste._

I don't want to complicate things if this is a niche problem and the possible solutions are problematic, but I am hoping there might be a clean way to get better results for roughness in the range [0.1, 0.2]. Suppose we can allow a single 1K texture for the environment, roughly scaling the current texture up to the nearest powers of two. Within that budget,<sup>1</sup> we could at least double the size of the mips associated with roughness values on this range, with either of the arrangements below:

![Screenshot 2023-09-19 at 10 56 01 PM](https://github.com/mrdoob/three.js/assets/1848368/06997ebe-d248-41d0-8299-324e8d53999e)

Please ignore my very arbitrary mip placement, for now. If I work out the mapping... does this sound like a reasonable thing to try?

**Comment by elalish** [comment]
You're welcome to try, but my sense is this will be a lot of work for a very minimal gain. Keep in mind that our PMREM already allows for variable size environment maps, which means if you know your minimum roughness is 0.15, then you can just upload a 64x128 image and get basically full quality with a tiny texture cache (and great framerate). Any difference to what resolution is supported at what roughness will also require increasing the blur sampling to avoid aliasing while keeping the proper sigma. You definitely want to maintain the steady power-of-two reduction in resolution or the filtering time will explode. 

The main reason I see this as a minimal gain is that with your example, if you just zoom in, those environment map pixels will get arbitrarily large, so while you might make one camera shot look better, others will still be highly pixelated. And in HDR, there's almost no amount of smoothing that will remove the square interpolation artifacts. You can see this problem clearly with roughness = 0 even if you upload a 4k environment map if you happen to have a flat, or worse yet, concave surface. There's just only so much we can do.

**Comment by sciecode** [comment]
I'm not 100% on this, but this could be a side-effect of `Roughness` remapping.
Some renderers use a very common remapping of roughness to a more user-friendly input `perceptualRoughness`

```js
roughness = perceptualRoughness * perceptualRoughness
```

[Filament](https://google.github.io/filament/Filament.html) has a section (4.8.3.3 - Roughness remapping and clamping ) on `perceptualRoughness`, which better explains the remapping. But the reasoning is that it better maps to what we perceive as roughness.

Top ( perceivedRoughness ) - Bottom ( Roughness )
![image](https://github.com/mrdoob/three.js/assets/3927951/2dd22ce4-3c43-40c1-9171-7ed667e41025)

I don't believe Three.js utilizes this remapping, so if instead of .15 roughness you utilize 0.0225, I think things would have a more direct mapping to what Khronos and PlayCanvas show.

**Comment by sciecode** [comment]
It appears I'm mistaken, either three.js already uses roughness remapping and I'm not aware, or differences are based on radianceMap generation.

THREE - top `perceptualRoughness` - bottom `roughness`
![image](https://github.com/mrdoob/three.js/assets/3927951/f088e72e-bf0c-4e44-83bf-0c86fc4bfb4a)

Khronos Sample Viewer:
![image](https://github.com/mrdoob/three.js/assets/3927951/4fdb9981-28ed-4b43-8737-4ad9b235f761)

**Comment by Alexander-103** [comment]
Чуть позже рассмотрю! 🔥

вс, 24 сент. 2023 г., 22:02 Guilherme Avila ***@***.***>:

**Comment by sciecode** [comment]
After digging a bit more, Three.js indeed already uses remapped roughness for it's normal workflow. Therefore the `roughness` parameter in all materials should already be considerered what `Filament` refers to as `perceptualRoughness`, I took this opportunity to make sure most shaders are being consistent in considering this remapping, and sure enough all PBR roughness usages I could find are indeed consistent:

https://github.com/mrdoob/three.js/blob/53c642e6905a1e53b0dd1c37ce14ec46f8b8fdc2/src/renderers/shaders/ShaderChunk/lights_physical_pars_fragment.glsl.js#L151
https://github.com/mrdoob/three.js/blob/53c642e6905a1e53b0dd1c37ce14ec46f8b8fdc2/src/renderers/shaders/ShaderChunk/envmap_physical_pars_fragment.glsl.js#L29

However the only place where I couldn't find this being considered is when sampling from `ENVMAP_CUBE_UV` using `textureCubeUV`. Internally the code directly uses material.roughness as it is, I can't confirm whether that is indeed the appropriate behaviour or not. Perhaps @elalish could confirm this.

https://github.com/mrdoob/three.js/blob/53c642e6905a1e53b0dd1c37ce14ec46f8b8fdc2/src/renderers/shaders/ShaderChunk/cube_uv_reflection_fragment.glsl.js#L168

**Comment by elalish** [comment]
Indeed, I went through pretty carefully to match up roughness values with the appropriate levels of filtering. You can read more about it in my [paper](https://drive.google.com/file/d/15y8r_UpKlU9SvV4ILb0C3qCPecS8pvLz/view).

**Comment by mrdoob** [comment]
@sciecode Did you manage to get anywhere with the investigation you were doing?

**Comment by sciecode** [comment]
I've reviewed our environment brdf implementation for any possible mistakes that might've slipped past, but everything looks in order. It's aligned both with academic literature and what other production renderers are doing at the moment. As far as what is being discussed in this topic, I wasn't able to find anything problematic.

On a somewhat related topic, I've noticed some visual inaccuracies with the usage of UE4 analytical BRDF split-sum approximation. 

https://github.com/mrdoob/three.js/blob/4527c3e8498ebdc42e7eb40fcfa90435b9c95514/src/renderers/shaders/ShaderChunk/lights_physical_pars_fragment.glsl.js#L372-L392

It's used as a substitution to the Environment BRDF LUT, mostly for mobile compatibility, as using a 16b float texture on mobile presents a series of problems.

I've been working on a slightly more accurate model that aims to improve on this approximation, I'll be prepping a PR in the near future, alongside with a blog-post that explains most of the reasoning and comparison between both methods. I've already seen a significant improvement in quality, but it bears no effect on the pixelated appearance that Don refers to, as the BRDF isn't cause of this issue.
